### DISCUSSION TITLE
Library using Pure.DI consumed by MS DI

### DISCUSSION THREAD

**Discussion by bstordrup** [DevTeam/Pure.DI#79] [Q&A]
If I create a library using Pure.DI for dependency injection, can it then be used in a project not using Pure.DI, but uses the traditional Dependency Injection from Microsoft?

**Comment by NikolayPianikov** [comment]
> If I create a library using Pure.DI for dependency injection, can it then be used in a project not using Pure.DI, but uses the traditional Dependency Injection from Microsoft?

Yes your library can of course be used with MS DI. 

The easiest way is to simply define some composition roots in your library, make them available from outside. A consumer of your library will be able to register them in the MS DI service collection using [extension methods](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.extensions.servicecollectiondescriptorextensions.tryaddtransient).

Another way, where you yourself will provide a collection of services from your library. [Here is an example](https://github.com/DevTeam/Pure.DI/blob/master/readme/service-collection.md) where all composition roots are added to the MS DI service collection automapically. In this case, you will need to add the [Pure.DI.MS](https://www.nuget.org/packages/Pure.DI.MS) package to your library.

You can also provide service providers as in the examples below, but I think this is less suitable for you.
- [Keyed service provider](https://github.com/DevTeam/Pure.DI/blob/master/readme/keyed-service-provider.md)
- [Service provider](https://github.com/DevTeam/Pure.DI/blob/master/readme/service-provider.md)
- [Service provider with scope](https://github.com/DevTeam/Pure.DI/blob/master/readme/service-provider-with-scope.md)

**Reply by bstordrup** [reply]
Hi, thank you for the reply. The link for the example with adding roots to MS ServiceCollection does not work.

**Reply by NikolayPianikov** [reply]
> Hi, thank you for the reply. The link for the example with adding roots to MS ServiceCollection does not work.

Sorry, I fixed the link

**Reply by NikolayPianikov** [reply]
Created [an example](https://github.com/user-attachments/files/18260367/ServiceCollectionExample.zip) for you

**Reply by NikolayPianikov** [reply]
But [this approach](https://github.com/user-attachments/files/18261109/ServiceCollectionExample.zip) is simpler and better
